File: AgentLens/Services/HomeAssistant/HomeAssistantConfig.swift
import Foundation

// MARK: - Home Assistant Recovery Config
//
// All non-secret state for the OpenBurnBar Smart Display recovery flow.
// The access token never lives here — it sits in `HomeAssistantTokenStore`.
// `webhookID` is the random secret OpenBurnBar provisioned when it
// installed the recovery automation; it is reproduced into the webhook
// URL on every recovery call.

struct HomeAssistantConfig: Codable, Equatable, Sendable {

    /// Normalized HA base URL ("http://homeassistant.local:8123",
    /// "https://my.example.com"). Does not contain trailing slash.
    var baseURL: URL

    /// Selected `media_player.*` entity that will be cast to during
    /// recovery. Empty until the user confirms the picker.
    var mediaPlayerEntityID: String

    /// Friendly display name surfaced in the UI / logs.
    var mediaPlayerFriendlyName: String

    /// Random URL-safe ID used as the webhook trigger ID inside HA.
    /// We prefix with `openburnbar_cast_recover_` to make it visually
    /// recognizable in users' HA configs.
    var webhookID: String

    /// HA-side automation entity ID we created or updated.
    var automationEntityID: String

    /// Whether the recovery automation is currently considered installed.
    var automationInstalled: Bool

    /// Whether the live test successfully reached HA at least once.
    var lastTestPassed: Bool

    /// ISO8601 timestamp of last successful verification.
    var lastVerifiedAt: Date?

    /// Mode used for setup. We branch UI on this in Settings.
    var setupMode: SetupMode

    enum SetupMode: String, Codable, Sendable, CaseIterable {
        case rest                 // Phase A — REST automation provisioning
        case blueprint            // Blueprint fallback path
        case manualWebhook        // User pasted a webhook URL by hand
    }

    init(
        baseURL: URL,
        mediaPlayerEntityID: String = "",
        mediaPlayerFriendlyName: String = "",
        webhookID: String = "",
        automationEntityID: String = "",
        automationInstalled: Bool = false,
        lastTestPassed: Bool = false,
        lastVerifiedAt: Date? = nil,
        setupMode: SetupMode = .rest
    ) {
        self.baseURL = baseURL
        self.mediaPlayerEntityID = mediaPlayerEntityID
        self.mediaPlayerFriendlyName = mediaPlayerFriendlyName
        self.webhookID = webhookID
        self.automationEntityID = automationEntityID
        self.automationInstalled = automationInstalled
        self.lastTestPassed = lastTestPassed
        self.lastVerifiedAt = lastVerifiedAt
        self.setupMode = setupMode
    }

    /// Webhook URL to call on the configured HA instance.
    /// HA exposes webhooks at `<baseURL>/api/webhook/<id>`.
    var webhookURL: URL? {
        guard !webhookID.isEmpty else { return nil }
        return baseURL.appendingPathComponent("api/webhook/\(webhookID)")
    }
}

// MARK: - URL Normalization

enum HomeAssistantURLNormalizer {

    /// Accepts free-form user input ("homeassistant.local",
    /// "http://homeassistant.local:8123/", "https://x.duckdns.org") and
    /// returns a canonical URL we can use as `baseURL`.
    ///
    /// Rules:
    ///   - If no scheme is present, we assume `http://` for `*.local`
    ///     hostnames and `https://` for everything else.
    ///   - We append `:8123` only for `http` schemes that omit a port,
    ///     since that's HA's default; HTTPS endpoints (Nabu Casa,
    ///     reverse proxies) typically already terminate on 443.
    ///   - We strip any trailing slash, query, and path.
    static func normalize(_ raw: String) -> URL? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        let withScheme: String
        if trimmed.contains("://") {
            withScheme = trimmed
        } else if trimmed.hasSuffix(".local") || trimmed.contains(".local:") {
            withScheme = "http://" + trimmed
        } else {
            withScheme = "https://" + trimmed
        }

        guard var components = URLComponents(string: withScheme),
              let host = components.host, !host.isEmpty else { return nil }

        // Drop path/query/fragment — we only want the origin.
        components.path = ""
        components.query = nil
        components.fragment = nil

        // Default port for plain http when missing.
        if components.scheme == "http", components.port == nil {
            components.port = 8123
        }

        return components.url
    }
}

File: windows/integrations/homeassistant/OpenBurnBar.Integrations.HomeAssistant/HomeAssistantConfig.cs
using System;
using System.Globalization;
using System.Text.Json.Serialization;

namespace OpenBurnBar.Integrations.HomeAssistant;

// Non-secret recovery config for the OpenBurnBar Smart Display flow.
//
// Parity: AgentLens/Services/HomeAssistant/HomeAssistantConfig.swift
//   struct HomeAssistantConfig + enum SetupMode + webhookURL.
//
// The access token never lives here (it sits in the token store); this is the
// JSON-serializable metadata the config store persists.

public enum HomeAssistantSetupMode
{
    /// Phase A — REST automation provisioning.
    Rest,

    /// Blueprint fallback path.
    Blueprint,

    /// User pasted a webhook URL by hand.
    ManualWebhook,
}

public sealed class HomeAssistantConfig : IEquatable<HomeAssistantConfig>
{
    /// Normalized HA base URL (no trailing slash).
    public string BaseUrl { get; set; }

    public string MediaPlayerEntityId { get; set; }

    public string MediaPlayerFriendlyName { get; set; }

    /// Random URL-safe webhook trigger id (prefixed openburnbar_cast_recover_).
    public string WebhookId { get; set; }

    public string AutomationEntityId { get; set; }

    public bool AutomationInstalled { get; set; }

    public bool LastTestPassed { get; set; }

    /// ISO8601 timestamp of last successful verification, or null.
    public DateTimeOffset? LastVerifiedAt { get; set; }

    public HomeAssistantSetupMode SetupMode { get; set; }

    public HomeAssistantConfig(
        string baseUrl,
        string mediaPlayerEntityId = "",
        string mediaPlayerFriendlyName = "",
        string webhookId = "",
        string automationEntityId = "",
        bool automationInstalled = false,
        bool lastTestPassed = false,
        DateTimeOffset? lastVerifiedAt = null,
        HomeAssistantSetupMode setupMode = HomeAssistantSetupMode.Rest)
    {
        BaseUrl = baseUrl;
        MediaPlayerEntityId = mediaPlayerEntityId;
        MediaPlayerFriendlyName = mediaPlayerFriendlyName;
        WebhookId = webhookId;
        AutomationEntityId = automationEntityId;
        AutomationInstalled = automationInstalled;
        LastTestPassed = lastTestPassed;
        LastVerifiedAt = lastVerifiedAt;
        SetupMode = setupMode;
    }

    /// Webhook URL on the configured HA instance: `<baseURL>/api/webhook/<id>`.
    /// Parity: Swift `webhookURL` (nil when no id).
    [JsonIgnore]
    public string? WebhookUrl
    {
        get
        {
            if (string.IsNullOrEmpty(WebhookId))
            {
                return null;
            }
            var trimmed = BaseUrl.EndsWith("/", StringComparison.Ordinal)
                ? BaseUrl.Substring(0, BaseUrl.Length - 1)
                : BaseUrl;
            return $"{trimmed}/api/webhook/{WebhookId}";
        }
    }

    public bool Equals(HomeAssistantConfig? other) =>
        other is not null &&
        BaseUrl == other.BaseUrl &&
        MediaPlayerEntityId == other.MediaPlayerEntityId &&
        MediaPlayerFriendlyName == other.MediaPlayerFriendlyName &&
        WebhookId == other.WebhookId &&
        AutomationEntityId == other.AutomationEntityId &&
        AutomationInstalled == other.AutomationInstalled &&
        LastTestPassed == other.LastTestPassed &&
        Nullable.Equals(LastVerifiedAt, other.LastVerifiedAt) &&
        SetupMode == other.SetupMode;

    public override bool Equals(object? obj) => Equals(obj as HomeAssistantConfig);

    public override int GetHashCode()
    {
        var hash = new HashCode();
        hash.Add(BaseUrl);
        hash.Add(MediaPlayerEntityId);
        hash.Add(MediaPlayerFriendlyName);
        hash.Add(WebhookId);
        hash.Add(AutomationEntityId);
        hash.Add(AutomationInstalled);
        hash.Add(LastTestPassed);
        hash.Add(LastVerifiedAt);
        hash.Add(SetupMode);
        return hash.ToHashCode();
    }
}

/// Raw-value mapping for HomeAssistantSetupMode matching the Swift enum's
/// string raw values ("rest" / "blueprint" / "manualWebhook").
public static class HomeAssistantSetupModeExtensions
{
    public static string RawValue(this HomeAssistantSetupMode mode) => mode switch
    {
        HomeAssistantSetupMode.Rest => "rest",
        HomeAssistantSetupMode.Blueprint => "blueprint",
        HomeAssistantSetupMode.ManualWebhook => "manualWebhook",
        _ => "rest",
    };

    public static HomeAssistantSetupMode ParseSetupMode(string? rawValue) => rawValue switch
    {
        "blueprint" => HomeAssistantSetupMode.Blueprint,
        "manualWebhook" => HomeAssistantSetupMode.ManualWebhook,
        _ => HomeAssistantSetupMode.Rest,
    };
}
